src/Vimeo/Vimeo.php

```diff
@@ -256,63 +256,92 @@ public function buildAuthorizationEndpoint($redirect_uri, $scope = 'public', $st
 
-        // Begin the upload request by getting a ticket
-        $ticket_args = array('type' => 'streaming');
-        if ($machine_id !== null) {
-            $ticket_args['machine_id'] = $machine_id;
+        $file_size = filesize($file_path);
+
+        // If the user does not have enough free space in their quota to upload this, then don't.
+        $response = $this->request('/me', array('fields' => 'upload_quota.space.free'), 'GET');
+        if ($response['status'] !== 200) {
+            $error = !empty($response['body']['error']) ? ' [' . $response['body']['error'] . ']' : '';
+            throw new VimeoUploadException('Unable to pull the users upload quota.' . $error);
+        } elseif ($file_size > $response['body']['upload_quota']['space']['free']) {
+            throw new VimeoUploadException('User does not have any more free space to upload this video.');
+        }
+
+        // Ignore any specified upload approach and size.
+        $params['upload']['approach'] = 'tus';
+        $params['upload']['size'] = $file_size;
+
+        // Use JSON filtering so we only receive the data that we need to make an upload happen.
+        $uri = '/me/videos?fields=uri,upload';
+
+        $attempt = $this->request($uri, $params, 'POST');
+        if ($attempt['status'] !== 200) {
+            $attempt_error = !empty($attempt['body']['error']) ? ' [' . $attempt['body']['error'] . ']' : '';
+            throw new VimeoUploadException('Unable to initiate an upload attempt.' . $attempt_error);
         }
-        $ticket = $this->request('/me/videos', $ticket_args, 'POST');
 
-        return $this->perform_upload($file_path, $ticket);
+        return $this->perform_upload_tus($file_path, $file_size, $attempt);
     }
 
     /**
      * Replace the source of a single Vimeo video.
      *
+     * @link https://developer.vimeo.com/api/endpoints/videos#PUT/videos/{video_id}/files
      * @param string $video_uri Video uri of the video file to replace.
      * @param string $file_path Path to the video file to upload.
-     * @param string|null $machine_id
      * @throws VimeoUploadException
      * @return string Status
      */
-    public function replace($video_uri, $file_path, $machine_id = null)
+    public function replace($video_uri, $file_path)
     {
         //  Validate that our file is real.
         if (!is_file($file_path)) {
             throw new VimeoUploadException('Unable to locate file to upload.');
         }
 
+        $file_size = filesize($file_path);
+
         $uri = $video_uri . self::REPLACE_ENDPOINT;
 
+        // Use JSON filtering so we only receive the data that we need to make an upload happen.
+        $uri .= '?fields=upload_link,complete_uri';
+
         // Begin the upload request by getting a ticket
-        $ticket_args = array('type' => 'streaming');
-        if ($machine_id !== null) {
-            $ticket_args['machine_id'] = $machine_id;
+        $params = array(
+            'type' => 'streaming'
```

Comment:
How come replace is the only functionality supporting `streaming`?

Reply:
We lock the SDKs to a single approach, and `tus` isn't supported on replace yet. If people want to use non-streaming, they can write their own integrations.